addons/mail/static/src/new/messaging_hook.js

```diff
@@ -36,3 +37,117 @@ export function useMessageHighlight(duration = 2000) {
     }
     return state;
 }
+
+function dataUrlToBlob(data, type) {
+    const binData = window.atob(data);
+    const uiArr = new Uint8Array(binData.length);
+    uiArr.forEach((_, index) => (uiArr[index] = binData.charCodeAt(index)));
+    return new Blob([uiArr], { type });
+}
+
+export function useAttachmentUploader({ threadId, messageId }) {
+    const { bus, upload } = useService("file_upload");
+    const notification = useService("notification");
+    const messaging = useService("mail.messaging");
+    let abortByUploadId = {};
+    const uploadingAttachmentIds = new Set();
+    const state = useState({
+        attachments: [],
+        async upload(fileData) {
+            const { name, data, type } = fileData;
+            const thread =
+                messaging.state.threads[threadId || messaging.state.messages[messageId].resId];
+            const file = new File([dataUrlToBlob(data, type)], name, { type });
+            const tmpId = messaging.nextId++;
+            uploadingAttachmentIds.add(tmpId);
+            upload("/mail/attachment/upload", [file], {
+                buildFormData(formData) {
+                    formData.append("thread_id", thread.resId || thread.id);
+                    formData.append("thread_model", thread.resModel || "mail.channel");
+                    formData.append("is_pending", true);
+                    formData.append("temporary_id", tmpId);
+                },
+            }).catch((e) => {
+                if (e.name !== "AbortError") {
+                    throw e;
+                }
+            });
+        },
+        async unlink(attachment) {
+            const abort = abortByUploadId[attachment.id];
+            if (abort) {
+                abort();
+                return;
+            }
+            await messaging.unlinkAttachment(attachment);
+            removeFromArrayWithPredicate(state.attachments, ({ id }) => id === attachment.id);
+        },
+        async unlinkAll() {
+            const proms = [];
+            this.attachments.forEach((attachment) => proms.push(this.unlink(attachment)));
+            await Promise.all(proms);
+            this.reset();
+        },
```

Comment:
I really dislike the terminology `unlink` for removing/deleting the record, even if it's the one chosen in python ORM, this sucks. There's ambiguity when we want to unlink without deletion...